046-analyse-broker-metrics-before-rolling-kafka.md

```diff
@@ -0,0 +1,86 @@
+- Configure the server to require TLS client authentication. 
+- Expose an endpoint `v1/brokermetrics` that includes the API version in the URI path.
+- Extend KafkaAgent to collect `kafka.log:type=LogManager,name=remainingLogsToRecover` and `kafka.log:type=LogManager,name=remainingSegmentsToRecover` metrics.
+- Take the metric name as a parameter in the request query e.g `v1/brokermetrics?name=BrokerState`. Use the value of the MBean's `name` field for the request  e.g. `BrokerState` and  `remainingLogsToRecover`.
+- Use Jetty's native ServerHandler API to handle a `GET` request and return 
+    - an integer value for `BrokerState`.
+    - an integer value for  `remainingLogsToRecover`.
+    - an integer value for `remainingSegmentsToRecover`.
+- If no metric name was provided in the request or the requested metric name is not supported, return HTTP 404 (Not Found) to the client.
+
+*KafkaRoller*:
+- Use the Pod DNS name to query the KafkaAgent API.
+- Use the cluster operator's client certificate and keys to authenticate with the KafkaAgent API server.
+- Instead of checking the broker's readiness through `isReady()` function, utilise the new RESTful endpoint to determine if the broker is ready.
+- Include an additional field in the status section of the Kafka CR. This will hold data that will be utilised by KafkaRoller in case the reconciliation process is repeated while the log recovery process is still ongoing:
+```
+status:
+    <kafka_pod_name>:
+      restartTimeStamp: 
+      logRecoveryExceededTimeout: <true|false>
+```
+- When KafkaRoller restarts a broker, write the creation timestamp to the new field, `restartTimeStamp`.
+- If a kafka pod does not become ready within the operational timeout, check the broker state by sending a request to the new endpoint:
+    - If log recovery is in progress, apply the new logic for handling log recovery.
+    - Otherwise, continue with the current behaviour.
+
+*Logic for handling log recovery*
+- Wait for log recovery to complete until the operation timeout (5 mins by default) reaches.
+- While waiting for log recovery to complete, check its progress by periodically sending a request to the new endpoint for `remainingLogsToRecover` and `remainingSegmentsToRecover` metrics and log them in an `INFO` message.
+- If the operation timeout is exceeded, check the value of `logRecoveryExceededTimeout` field in the CR:
+    - If false, log a `WARN` message, fail the reconciliation and set `logRecoveryExceededTimeout` to true.
+    - If true, fail the reconciliation.
+- If kafka pod becomes ready, check the value of `logRecoveryExceededTimeout`. If set to true:
+    - Calculate the time passed since the `restartTimeStamp` and log it in a WARN message. This would be helpful to figure out a more appropriate timeout value to avoid this in the future.
+    - Empty `restartTimeStamp` and `logRecoveryExceededTimeout` values.
```

Comment:
TBH, I do not get this. Can you elaborate on why this is useful? Why not keep it simple and if your wait for the broker to be ready is over and it is still recovering, just raise a WARN log message and be done with it. Or maybe I did not understand properly what exactly is the added value of this mechanism.

Reply:
Now I see this written down it looks pretty convoluted. What we were trying to achieve here was a mechanism to allow the user to know how long their brokers are _actually_ take to recover. On reflection they can get that easily enough from the broker metrics themselves (e.g. from prometheus/grafana). So there's no great value in futzing around trying to log something (whether once or many times): We should just fail the reconciliation and let the user figure it out. That would allow us to keep the simpler endpoint for the broker state metrics (returning a single integer), keep this stuff out of the CR `status` entirely and the user would still be able to reconfigure the CR (in the event that log recovery was really taking a long time).

Sorry for having led you astray here @tinaselenge.

Reply:
I agree that this is unnecessary complexity and simply failing the reconciliation with a message that shows log recover progress is enough.

Reply:
I agree with this. Reconciliation failure already gets logged with a cause, so thinking about it again , we probably won't need this conditional WARN log. Also like you said, user can monitor the metrics for log recovery progress directly from the brokers and it's better to see the metric trend over time, rather than looking at one event from log. I will go ahead remove the parts about writing to the CR.